include/xtensor/xhistogram.hpp

```diff
@@ -57,46 +57,94 @@ namespace xt
     template <class R = double, class E1, class E2, class E3>
     inline auto histogram(E1&& data, E2&& bin_edges, E3&& weights, bool density = false)
     {
-        using size_type = common_size_type_t<std::decay_t<E1>, std::decay_t<E2>, std::decay_t<E3>>;
-        using value_type = typename std::decay_t<E3>::value_type;
-
-        XTENSOR_ASSERT(data.dimension() == 1);
-        XTENSOR_ASSERT(weights.dimension() == 1);
-        XTENSOR_ASSERT(bin_edges.dimension() == 1);
-        XTENSOR_ASSERT(weights.size() == data.size());
-        XTENSOR_ASSERT(bin_edges.size() >= 2);
-        XTENSOR_ASSERT(std::is_sorted(bin_edges.cbegin(), bin_edges.cend()));
-        XTENSOR_ASSERT(xt::amin(data)[0] >= bin_edges[0]);
-        XTENSOR_ASSERT(xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1]);
-
-        xt::xtensor<value_type, 1> count = xt::zeros<value_type>({ bin_edges.size() - 1 });
-
-        auto sorter = xt::argsort(data);
-
-        size_type ibin = 0;
+        return histogramImp(std::forward<E1>(data),
+                            std::forward<E2>(data),
+                            std::forward<E3>(weights),
+                            density,
+                            false);
+    }
 
-        for (auto& idx : sorter)
+    namespace detail
+    {
+        /**
+         * @ingroup histogram
+         * @brief Compute the histogram of a set of data.
+         *
+         * @param data The data.
+         * @param bin_edges The bin-edges. It has to be 1-dimensional and monotonic.
+         * @param weights Weight factors corresponding to each data-point.
+         * @param density If true the resulting integral is normalized to 1. [default: false]
+         * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+         */
+        template <class R = double, class E1, class E2, class E3>
+        inline auto histogramImp(E1&& data, E2&& bin_edges, E3&& weights, bool density = false, bool equal_bins=false)
```

Comment:
> Formatting: we use snake case, so this method should b renamed into historgram_impl

Done.

> We don't document detail or implementation methods since they're not part of the public API

Removed.

> Since this function is supposed to be called from public API, all the arguments are supposed to be passed, so no need to provide default values.

Default values were removed.

> equal_bins is known at compile time, so maybe tag dispatching could be use instead of a boolean parameter? This would simplify the code

Here I am not so sure. By using tag dispatch, we will have two `histogram_imp` and a lot of code (the `XTENSOR_ASSERT`s and the last 12 lines of the function) will be duplicated. Of course, we can have two small functions to implement the code in `if (equal_bins) { ... } else { ... }`, respectively.